### PR TITLE
Python: Upgrade to Python 3.8 as the "default" Python version

### PR DESCRIPTION
**Pull request by Harshitha1234**
Signed-off-by: Harshitha1234 <harshithachowdary.t17@iiits.in>

**Changes**
Upgrade to Python 3.8 [Bug 1624680](https://bugzilla.mozilla.org/show_bug.cgi?id=1624680)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -897,7 +897,7 @@ jobs:
       - run:
           name: Setup default Python version
           command: |
-            echo "export PATH=/opt/python/cp37-cp37m/bin:$PATH" >> $BASH_ENV
```
Comment (Harshitha1234) [inline .circleci/config.yml:900]: Does this line require changes? :)

Comment (mdboom) [reply .circleci/config.yml:900]: Yes, I think so.

**Comment by mdboom** [comment]
It looks like the "Generate Python documentation" job is still using a Python 3.7 container.

**Comment by mdboom** [comment]
The change in the behavior of `locale.getlocale()` is really surprising -- I can't see anything in the Python changelog that would result in that change.  We'll want to fix that, since Glean should be sending locales in Unix-style RFC 1766 format, not in the longer Windows-style format that Python 3.8 seems to be returning.  (I'm looking at that issue and will report back.)

**Comment by Harshitha1234** [comment]
Great! Would wait till then 👍

**Comment by Dexterp37** [comment]
See also #1004

**Comment by mdboom** [comment]
#1004 should have fixed the locale test failure.  @Harshitha1234 : Would you mind rebasing this PR on the `main` branch so we could try again?  (Let me know if you need help with that).

**Comment by Harshitha1234** [comment]
My bad! I have to rebase it properly :)

**Comment by Harshitha1234** [comment]
Any thoughts or ideas on how to deal with the documentation check?

**Comment by Harshitha1234** [comment]
> Thanks for all this work! It looks like the one remaining issue is the container being used for "Generate Python documentation".

Cool! Any thoughts on how to fix it? 🙂

**Comment by mdboom** [comment]
The documentation job is using 3.7 for the container, but the path is updated to 3.8.  Both things should be at 3.8.

**Comment by Dexterp37** [comment]
Hey @Harshitha1234 , thank you for this initial PR and for contributing to the project! This is almost finished and just requires a few tweaks. Are you planning on coming back to this?

**Comment by Harshitha1234** [comment]
Yeah, I would love to. I've few time commitments till 5th July :) Can I look into this on 5th? Thanks :)

**Comment by Dexterp37** [comment]
> 
> 
> Yeah, I would love to. I've few time commitments till 5th July :) Can I look into this on 5th? Thanks :)

Sure! I just wanted to check-in to see if you needed support or anything :-)

File: .circleci/config.yml [hunk 2]
```diff
@@ -872,14 +872,14 @@ jobs:
...
-      - image: circleci/python:3.7.6
+      - image: circleci/python:3.8.2
     steps:
       - checkout
       - attach_workspace:
           at: glean-core/python/
       - run:
           name: Generate Python docs
-          command: glean-core/python/.venv3.7/bin/python3 -m pdoc --html glean --force -o build/docs/python
+          command: glean-core/python/.venv3.8/bin/python3 -m pdoc --html glean --force -o build/docs/python
```
Comment (mdboom) [inline .circleci/config.yml:882]: Let's try this:

```suggestion
          command: glean-core/python/.venv3.8/bin/python -m pdoc --html glean --force -o build/docs/python
```

**Comment by mdboom** [comment]
Hmm... This is a tricky one.  I think the issue is that the documentation job is still depending on the Python 3.7 tests here:

https://github.com/mozilla/glean/blob/main/.circleci/config.yml#L1046

**Comment by Harshitha1234** [comment]
> Hmm... This is a tricky one. I think the issue is that the documentation job is still depending on the Python 3.7 tests here:
> 
> https://github.com/mozilla/glean/blob/main/.circleci/config.yml#L1046

Replace it with Python 3_8 tests?

**Comment by mdboom** [comment]
> > Hmm... This is a tricky one. I think the issue is that the documentation job is still depending on the Python 3.7 tests here:
> > https://github.com/mozilla/glean/blob/main/.circleci/config.yml#L1046
> 
> Replace it with Python 3_8 tests?

Yes

**Comment by mdboom** [comment]
> > > Hmm... This is a tricky one. I think the issue is that the documentation job is still depending on the Python 3.7 tests here:
> > > https://github.com/mozilla/glean/blob/main/.circleci/config.yml#L1046
> > 
> > 
> > Replace it with Python 3_8 tests?
> 
> Yes

And there are a bunch of analogous changes required in the release jobs -- all the Python 3.7 stuff referenced there should also be moved to 3.8

**Comment by mdboom** [comment]
Thanks.  It looks like you need to rebase now to resolve the conflicts.

**Comment by mdboom** [comment]
Sorry this is a rabbit hole ;)  The Python 3.8 test is not persisting to the workspace (like the Python 3.7 test does), so the files aren't available when it tries to run the Python documentation job.

**Comment by Harshitha1234** [comment]
> Sorry this is a rabbit hole ;) The Python 3.8 test is not persisting to the workspace (like the Python 3.7 test does), so the files aren't available when it tries to run the Python documentation job.

So, what would be the fix? :)

**Comment by mdboom** [comment]
The Python 3.8 tests need a `persist_to_workspace` clause so that the files it generates will be available to subsequent jobs (in this case, the Generate Python documentation job).

**Comment by Harshitha1234** [comment]
Finally!! It passed the python documentation check 😭 I'm really happy and apologies for the delay

> The Python 3.8 tests need a `persist_to_workspace` clause so that the files it generates will be available to subsequent jobs (in this case, the Generate Python documentation job).

**Comment by Harshitha1234** [comment]
Never imagined contributing to Mozilla 🙂 and this one's my first PR in Mozilla. I'm so happy and thank you for the constant support and guidance at every step

**Comment by travis79** [comment]
> Never imagined contributing to Mozilla 🙂 and this one's my first PR in Mozilla. I'm so happy and thank you for the constant support and guidance at every step

Congratulations!  We very much appreciate your contribution and hope you consider contributing more!

**Comment by Harshitha1234** [comment]
Is Bugzilla, the main platform to find issues? 🙂

**Comment by travis79** [comment]
> Is Bugzilla, the main platform for finding issues? 🙂

You can use Bugzilla, but we have a dashboard that might be easier to use: https://dexterp37.github.io/bugzhub/#glean

**Comment by mdboom** [comment]
As Travis said, thanks for the contribution!  I'm going to hit that merge button now.

**Pull request merged**